solr/src/main/scala/akka/stream/alpakka/solr/SolrFlowStage.scala

```diff
@@ -117,60 +228,129 @@ private[solr] final class SolrFlowLogic[T, C](
   }
 
   private def tryPull(): Unit =
-    if (queue.size < settings.bufferSize && !isClosed(in) && !hasBeenPulled(in)) {
+    if (!isClosed(in) && !hasBeenPulled(in)) {
       pull(in)
     }
 
-  private def handleFailure(messages: Seq[IncomingMessage[T, C]], exc: Throwable): Unit =
+  private def handleFailure(args: (Seq[IncomingMessage[T, C]], Throwable)): Unit = {
+    val (messages, exc) = args
     if (retryCount >= settings.maxRetry) {
+      log.warning(s"Received error from solr. Giving up after $retryCount tries. Error: ${exc.toString}")
       failStage(exc)
     } else {
       retryCount = retryCount + 1
+      log.warning(
+        s"Received error from solr. (re)tryCount: $retryCount maxTries: ${settings.maxRetry}. Error: ${exc.toString}"
+      )
       failedMessages = messages
       scheduleOnce(NotUsed, settings.retryInterval)
     }
+  }
 
-  private def handleResponse(messages: Seq[IncomingMessage[T, C]], status: Int): Unit = {
+  private def handleResponse(args: (Seq[IncomingMessage[T, C]], Int)): Unit = {
+    val (messages, status) = args
+    log.debug(s"Handle the response with $status")
     retryCount = 0
-    val result = messages.map(m => IncomingMessageResult(m.source, m.passThrough, status))
+    val result = messages.map(m => IncomingMessageResult(m.idOpt, m.sourceOpt, m.passThrough, status))
 
-    emit(out, Future.successful(result))
+    emit(out, result)
 
-    val nextMessages = (1 to settings.bufferSize).flatMap { _ =>
-      queue.dequeueFirst(_ => true)
+    state match {
+      case Finished => handleSuccess()
+      case _ => state = Idle
     }
+  }
 
-    if (nextMessages.isEmpty) {
-      state match {
-        case Finished => handleSuccess()
-        case _ => state = Idle
+  private def handleSuccess(): Unit =
+    completeStage()
+
+  private def updateBulkToSolr(messages: Seq[IncomingMessage[T, C]]): UpdateResponse = {
+    val binder = messageBinder.get
```

Comment:
This might fail with a very unexpected error. Fail with a message users might understand, please.